src/bvh/collision_object/impl.hpp

```diff
@@ -109,11 +110,16 @@ namespace bvh
       logger.debug( "obj={} sending narrowphase patch {} with {} num elements",
                     collision_idx, vt_index{ _local_idx + rank * overdecomposition }, nelements );
       // Should be replaced with VT serialization
+
+      auto h_split_indices = Kokkos::create_mirror_view_and_copy( bvh::host_execution_space{}, split_indices );
+      auto h_data = Kokkos::create_mirror_view_and_copy( bvh::host_execution_space{}, m_entity_ptr );
       for (std::size_t j = sbeg; j < send; ++j)
       {
         debug_assert( offset < send_msg->data_size, "split index offset={} is out of bounds (local data size is {})", offset, send_msg->data_size );
-        debug_assert( split_indices_h( j ) < snapshots.extent( 0 ), "user index is out of bounds" );
-        std::memcpy( &send_msg->user_data()[offset], m_entity_ptr + (split_indices_h( j ) * m_entity_unit_size), m_entity_unit_size);
+        debug_assert( h_split_indices( j ) < snapshots.extent( 0 ), "user index is out of bounds" );
+        // FIXME_CUDA: use subviews, deep_copy to host
+        std::memcpy( &send_msg->user_data()[offset], h_data.data() + ( h_split_indices( j ) * m_entity_unit_size ),
+                     m_entity_unit_size );
```

Comment:
The idea here is to avoid calling `deep_copy` inside the loop and use host mirrors of entire views for improved performance :thinking:
Still getting segmentation fault though, so this definitely needs more work.

Reply:
`Thread 1 "BVHTests" received signal SIGSEGV, Segmentation fault.`

```cpp
(cuda-gdb) bt
#0  0x00007ffff456feb0 in ?? () from /lib/x86_64-linux-gnu/libcuda.so.1
#1  0x00007ffff461a27a in ?? () from /lib/x86_64-linux-gnu/libcuda.so.1
#2  0x00007ffff48b5540 in ?? () from /lib/x86_64-linux-gnu/libcuda.so.1
#3  0x00007ffff451491e in ?? () from /lib/x86_64-linux-gnu/libcuda.so.1
#4  0x00007ffff488f040 in ?? () from /lib/x86_64-linux-gnu/libcuda.so.1
#5  0x00007ffff450dbd2 in ?? () from /lib/x86_64-linux-gnu/libcuda.so.1
#6  0x00007ffff46d6d9a in ?? () from /lib/x86_64-linux-gnu/libcuda.so.1
#7  0x00007ffff6e41135 in ?? ()
   from /usr/local/cuda/targets/x86_64-linux/lib/libcudart.so.11.0
#8  0x00007ffff6e16905 in ?? ()
   from /usr/local/cuda/targets/x86_64-linux/lib/libcudart.so.11.0
#9  0x00007ffff6e631c5 in cudaMemcpy ()
   from /usr/local/cuda/targets/x86_64-linux/lib/libcudart.so.11.0
#10 0x00007ffff6b496bd in Kokkos::Impl::CudaInternal::cuda_memcpy_wrapper<true> (
    this=0x7ffff6ca5c40 <Kokkos::Impl::CudaInternal::singleton()::self>, 
    dst=0x55555a3fb540, src=0x7fffc9c93080, count=472, kind=cudaMemcpyDefault)
    at /tmp/root/spack-stage/spack-stage-kokkos-4.4.01-goqq2ugj5eigruowfhbxsndg6xmdr6ta/spack-src/core/src/Cuda/Kokkos_Cuda_Instance.hpp:365
#11 0x00007ffff6b45403 in Kokkos::Impl::DeepCopyCuda (dst=0x55555a3fb540, 
    src=0x7fffc9c93080, n=472)
    at /tmp/root/spack-stage/spack-stage-kokkos-4.4.01-goqq2ugj5eigruowfhbxsndg6xmdr6ta/spack-src/core/src/Cuda/Kokkos_CudaSpace.cpp:62
#12 0x00005555567fe677 in Kokkos::Impl::DeepCopy<Kokkos::HostSpace, Kokkos::CudaSpace, Kokkos::OpenMP, void>::DeepCopy (this=0x7fffffffbcc0, dst=0x55555a3fb540, src=0x7fffc9c93080, 
    n=472) at /opt/view/include/Cuda/Kokkos_CudaSpace.hpp:554
#13 0x000055555698b8e7 in Kokkos::deep_copy<std::byte*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::Experimental::EmptyViewHooks, Kokkos::MemoryTraits<0u>, std::byte const*, Kokkos::LayoutLeft, Kokkos::Cuda, Kokkos::MemoryTraits<1u> >(Kokkos::View<std::byte*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::Experimental::EmptyViewHooks, Kokkos::MemoryTraits<0u> > const&, Kokkos::View<std::byte const*, Kokkos::LayoutLeft, Kokkos::Cuda, Kokkos::MemoryTraits<1u> > const&, std::enable_if<(std::is_void<Kokkos::ViewTraits<std::byte*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::Experimental::EmptyViewHooks, Kokkos::MemoryTraits<0u> >::specialize>::value&&std::is_void<Kokkos::ViewTraits<std::byte const*, Kokkos::LayoutLeft, Kokkos::Cuda, Kokkos::MemoryTraits<1u> >::specialize>::value)&&((((unsigned int)Kokkos::ViewTraits<std::byte*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::Experimental::EmptyViewHooks, Kokkos::MemoryTraits<0u> >::rank)!=(0))||(((unsigned int)Kokkos::ViewTraits<std::byte const*, Kokkos::LayoutLeft, Kokkos::Cuda, Kokkos::MemoryTraits<1u> >::rank)!=(0))), void>::type*) (dst=..., src=...) at /opt/view/include/Kokkos_CopyViews.hpp:1794
#14 0x000055555697bbc7 in Kokkos::create_mirror_view_and_copy<Kokkos::HostSpace, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::byte const*, Kokkos::LayoutLeft, Kokkos::Cuda, Kokkos::MemoryTraits<1u>, void> (arg_prop=..., src=...) at /opt/view/include/Kokkos_CopyViews.hpp:3769
#15 0x000055555697bd16 in Kokkos::create_mirror_view_and_copy<Kokkos::OpenMP, std::byte const*, Kokkos::LayoutLeft, Kokkos::Cuda, Kokkos::MemoryTraits<1u>, void> (src=..., name=...) at /opt/view/include/Kokkos_CopyViews.hpp:3790
#16 0x0000555556978c23 in bvh::collision_object::impl::prepare_local_patch_for_sending (this=0x55555a3ed680, _local_idx=1, _rank=0) at /distBVH/src/bvh/collision_object/impl.hpp:115

```